File: apps/wasm/static/shared/signaling-client.js
// WebSocket Signaling Client Module
// Manages WebSocket connection to Go signaling server for WebRTC setup

/**
 * Connection states for the signaling client
 */
export const SignalingState = {
    DISCONNECTED: 'disconnected',
    CONNECTING: 'connecting',
    CONNECTED: 'connected',
    RECONNECTING: 'reconnecting'
};

/**
 * Simple event emitter for signaling events
 */
class EventEmitter {
    constructor() {
        this._listeners = {};
    }

    on(event, callback) {
        if (!this._listeners[event]) {
            this._listeners[event] = [];
        }
        this._listeners[event].push(callback);
    }

    off(event, callback) {
        if (!this._listeners[event]) return;
        this._listeners[event] = this._listeners[event].filter(cb => cb !== callback);
    }

    emit(event, ...args) {
        if (!this._listeners[event]) return;
        this._listeners[event].forEach(callback => {
            try {
                callback(...args);
            } catch (err) {
                console.error(`Error in signaling event listener for ${event}:`, err);
            }
        });
    }
}

/**
 * WebSocket signaling client for WebRTC peer-to-peer connection setup
 */
export class SignalingClient {
    /**
     * @param {Object} options - Configuration options
     * @param {string} [options.url] - WebSocket URL (default: derived from window.location)
     * @param {number} [options.reconnectDelay] - Initial reconnect delay in ms (default: 1000)
     * @param {number} [options.maxReconnectDelay] - Maximum reconnect delay in ms (default: 30000)
     * @param {number} [options.reconnectMultiplier] - Delay multiplier for exponential backoff (default: 1.5)
     * @param {number} [options.maxReconnectAttempts] - Maximum reconnect attempts (default: 10)
     */
    constructor(options = {}) {
        this._url = options.url || null;
        this._reconnectDelay = options.reconnectDelay || 1000;
        this._maxReconnectDelay = options.maxReconnectDelay || 30000;
        this._reconnectMultiplier = options.reconnectMultiplier || 1.5;
        this._maxReconnectAttempts = options.maxReconnectAttempts || 10;

        this._ws = null;
        this._roomId = null;
        this._peerId = null;
        this._state = SignalingState.DISCONNECTED;
        this._currentReconnectDelay = this._reconnectDelay;
        this._reconnectAttempts = 0;
        this._reconnectTimer = null;
        this._shouldReconnect = false;
        this._emitter = new EventEmitter();
    }

    /**
     * Get current connection state
     * @returns {string}
     */
    get state() {
        return this._state;
    }

    /**
     * Get current room ID
     * @returns {string|null}
     */
    get roomId() {
        return this._roomId;
    }

    /**
     * Get local peer ID
     * @returns {string|null}
     */
    get peerId() {
        return this._peerId;
    }

    /**
     * Derive WebSocket URL from current page location
     * @returns {string}
     * @private
     */
    _deriveUrl() {
        if (this._url) return this._url;

        const protocol = window.location.protocol === 'https:' ? 'wss:' : 'ws:';
        const host = window.location.host;
        return `${protocol}//${host}/ws`;
    }

    /**
     * Connect to the signaling server and join a room
     * @param {string} roomId - Room ID to join
     * @param {string} peerId - Local peer ID
     * @returns {Promise<void>}
     */
    connect(roomId, peerId) {
        return new Promise((resolve, reject) => {
            if (this._ws && this._ws.readyState === WebSocket.OPEN) {
                // Already connected, leave current room and join new one
                this._roomId = roomId;
                this._peerId = peerId;
                this._sendJoin();
                resolve();
                return;
            }

            this._roomId = roomId;
            this._peerId = peerId;
            this._shouldReconnect = true;
            this._state = SignalingState.CONNECTING;
            this._emitter.emit('statechange', this._state);

            const url = this._deriveUrl();
            try {
                this._ws = new WebSocket(url);
            } catch (err) {
                this._state = SignalingState.DISCONNECTED;
                this._emitter.emit('statechange', this._state);
                reject(err);
                return;
            }

            const onOpen = () => {
                cleanup();
                this._state = SignalingState.CONNECTED;
                this._currentReconnectDelay = this._reconnectDelay;
                this._reconnectAttempts = 0;
                this._emitter.emit('statechange', this._state);
                this._setupMessageHandler();
                this._sendJoin();
                resolve();
            };

            const onError = (err) => {
                cleanup();
                this._state = SignalingState.DISCONNECTED;
                this._emitter.emit('statechange', this._state);
                reject(err);
            };

            const onClose = () => {
                cleanup();
                this._state = SignalingState.DISCONNECTED;
                this._emitter.emit('statechange', this._state);
                reject(new Error('WebSocket closed before connection established'));
            };

            const cleanup = () => {
                this._ws.removeEventListener('open', onOpen);
                this._ws.removeEventListener('error', onError);
                this._ws.removeEventListener('close', onClose);
            };

            this._ws.addEventListener('open', onOpen);
            this._ws.addEventListener('error', onError);
            this._ws.addEventListener('close', onClose);
        });
    }

    /**
     * Set up WebSocket message and close handlers
     * @private
     */
    _setupMessageHandler() {
        this._ws.onmessage = (event) => {
            try {
                const message = JSON.parse(event.data);
                this._handleMessage(message);
            } catch (err) {
                console.error('Error parsing signaling message:', err);
            }
        };

        this._ws.onclose = () => {
            const wasConnected = this._state === SignalingState.CONNECTED;
            this._state = SignalingState.DISCONNECTED;
            this._emitter.emit('statechange', this._state);
            this._emitter.emit('disconnected');

            if (wasConnected && this._shouldReconnect) {
                this._scheduleReconnect();
            }
        };

        this._ws.onerror = (err) => {
            this._emitter.emit('error', err);
        };
    }

    /**
     * Handle incoming signaling message
     * @param {Object} message
     * @private
     */
    _handleMessage(message) {
        switch (message.type) {
            case 'joined':
                // Confirmation that we joined the room
                this._emitter.emit('joined', message.room, message.peers || []);
                break;

            case 'peer-joined':
                // Another peer joined the room
                this._emitter.emit('peerjoined', message.peer_id);
                break;

            case 'peer-left':
                // A peer left the room
                this._emitter.emit('peerleft', message.peer_id);
                break;

            case 'offer':
                // Received SDP offer from another peer
                this._emitter.emit('offer', message.from, message.sdp);
                break;

            case 'answer':
                // Received SDP answer from another peer
                this._emitter.emit('answer', message.from, message.sdp);
                break;

            case 'ice-candidate':
                // Received ICE candidate from another peer
                this._emitter.emit('icecandidate', message.from, message.candidate);
                break;

            case 'error':
                // Server error message
                this._emitter.emit('servererror', message.message);
                break;

            default:
                console.warn('Unknown signaling message type:', message.type);
        }
    }

    /**
     * Send join message to server
     * @private
     */
    _sendJoin() {
        this._send({
            type: 'join',
            room: this._roomId,
            peer_id: this._peerId
        });
    }

    /**
     * Send a message through WebSocket
     * @param {Object} message
     * @private
     */
    _send(message) {
        if (this._ws && this._ws.readyState === WebSocket.OPEN) {
            this._ws.send(JSON.stringify(message));
        }
    }

    /**
     * Schedule a reconnection attempt with exponential backoff
     * @private
     */
    _scheduleReconnect() {
        if (this._reconnectAttempts >= this._maxReconnectAttempts) {
            this._emitter.emit('maxreconnectsreached');
            return;
        }

        this._state = SignalingState.RECONNECTING;
        this._emitter.emit('statechange', this._state);
        this._emitter.emit('reconnecting', this._reconnectAttempts + 1, this._currentReconnectDelay);

        this._reconnectTimer = setTimeout(() => {
            this._reconnectAttempts++;
            this._attemptReconnect();
        }, this._currentReconnectDelay);

        // Exponential backoff with jitter
        const jitter = Math.random() * 0.3 + 0.85; // 0.85-1.15
        this._currentReconnectDelay = Math.min(
            this._currentReconnectDelay * this._reconnectMultiplier * jitter,
            this._maxReconnectDelay
        );
    }

    /**
     * Attempt to reconnect
     * @private
     */
    _attemptReconnect() {
        if (!this._shouldReconnect) return;

        this.connect(this._roomId, this._peerId)
            .then(() => {
                this._emitter.emit('reconnected');
            })
            .catch((err) => {
                console.error('Reconnect attempt failed:', err);
                if (this._shouldReconnect) {
                    this._scheduleReconnect();
                }
            });
    }

    /**
     * Send an SDP offer to a specific peer
     * @param {string} targetPeerId - Target peer ID
     * @param {RTCSessionDescriptionInit} sdp - SDP offer
     */
    sendOffer(targetPeerId, sdp) {
        this._send({
            type: 'offer',
            target: targetPeerId,
            sdp: sdp
        });
    }

    /**
     * Send an SDP answer to a specific peer
     * @param {string} targetPeerId - Target peer ID
     * @param {RTCSessionDescriptionInit} sdp - SDP answer
     */
    sendAnswer(targetPeerId, sdp) {
        this._send({
            type: 'answer',
            target: targetPeerId,
            sdp: sdp
        });
    }

    /**
     * Send an ICE candidate to a specific peer
     * @param {string} targetPeerId - Target peer ID
     * @param {RTCIceCandidateInit} candidate - ICE candidate
     */
    sendIceCandidate(targetPeerId, candidate) {
        this._send({
            type: 'ice-candidate',
            target: targetPeerId,
            candidate: candidate
        });
    }

    /**
     * Leave the current room
     */
    leave() {
        if (this._ws && this._ws.readyState === WebSocket.OPEN) {
            this._send({
                type: 'leave',
                room: this._roomId,
                peer_id: this._peerId
            });
        }
        this._roomId = null;
    }

    /**
     * Disconnect from the signaling server
     */
    disconnect() {
        this._shouldReconnect = false;
        if (this._reconnectTimer) {
            clearTimeout(this._reconnectTimer);
            this._reconnectTimer = null;
        }
        if (this._ws) {
            this._ws.close();
            this._ws = null;
        }
        this._state = SignalingState.DISCONNECTED;
        this._emitter.emit('statechange', this._state);
    }

    /**
     * Check if currently connected
     * @returns {boolean}
     */
    isConnected() {
        return this._ws && this._ws.readyState === WebSocket.OPEN;
    }

    /**
     * Register an event listener
     * @param {string} event - Event name
     * @param {Function} callback - Callback function
     */
    on(event, callback) {
        this._emitter.on(event, callback);
    }

    /**
     * Remove an event listener
     * @param {string} event - Event name
     * @param {Function} callback - Callback function
     */
    off(event, callback) {
        this._emitter.off(event, callback);
    }
}

File: core/net/common/SignalingClient.cc
// SignalingClient implementation
// Uses WSConnection for WebSocket transport

#include "core/net/include/SignalingClient.h"

#include <random>

#include "core/log/include/Logger.h"
#include "core/net/include/SignalingProtocol.h"

namespace cells::net {

const char* signalingClientStateToString(SignalingClientState state) {
    switch (state) {
        case SignalingClientState::DISCONNECTED:
            return "DISCONNECTED";
        case SignalingClientState::CONNECTING:
            return "CONNECTING";
        case SignalingClientState::CONNECTED:
            return "CONNECTED";
        case SignalingClientState::RECONNECTING:
            return "RECONNECTING";
        case SignalingClientState::IN_ROOM:
            return "IN_ROOM";
    }
    return "UNKNOWN";
}

SignalingClient::SignalingClient(SignalingClientConfig config)
    : config_(std::move(config)), current_reconnect_delay_ms_(config_.reconnect_delay_ms) {}

SignalingClient::~SignalingClient() {
    cancelReconnect();
    if (ws_) {
        ws_->setDelegate(nullptr);
        ws_->close();
    }
}

void SignalingClient::connect(const std::string& room_id, const std::string& peer_id) {
    room_id_ = room_id;
    peer_id_ = peer_id;
    should_reconnect_ = true;

    LOG_INFO("[Signaling] Connecting to %s", config_.url.c_str());

    // If already connected, just join the new room
    if (ws_ && ws_->isConnected()) {
        sendJoin();
        return;
    }

    setState(SignalingClientState::CONNECTING);

    // Create WebSocket connection
    ws_ = WSConnection::make(config_.url);
    if (!ws_) {
        LOG_ERROR("[Signaling] Failed to create WebSocket");
        setState(SignalingClientState::DISCONNECTED);
        return;
    }

    ws_->setDelegate(this);
    ws_->connect();
}

void SignalingClient::disconnect() {
    should_reconnect_ = false;
    cancelReconnect();

    if (ws_) {
        // Send leave message if in a room
        if (!room_id_.empty() && ws_->isConnected()) {
            sendMessage(SignalingProtocol::buildLeaveMessage(room_id_, peer_id_));
        }

        ws_->setDelegate(nullptr);
        ws_->close();
        ws_.reset();
    }

    room_id_.clear();
    setState(SignalingClientState::DISCONNECTED);
}

void SignalingClient::leaveRoom() {
    if (!room_id_.empty() && ws_ && ws_->isConnected()) {
        sendMessage(SignalingProtocol::buildLeaveMessage(room_id_, peer_id_));
    }
    room_id_.clear();

    if (state_ == SignalingClientState::IN_ROOM) {
        setState(SignalingClientState::CONNECTED);
    }
}

void SignalingClient::sendOffer(const std::string& target_peer, const SessionDescription& sdp) {
    if (ws_ && ws_->isConnected()) {
        sendMessage(SignalingProtocol::buildOfferMessage(target_peer, sdp));
    }
}

void SignalingClient::sendAnswer(const std::string& target_peer, const SessionDescription& sdp) {
    if (ws_ && ws_->isConnected()) {
        sendMessage(SignalingProtocol::buildAnswerMessage(target_peer, sdp));
    }
}

void SignalingClient::sendICECandidate(const std::string& target_peer,
                                       const ICECandidate& candidate) {
    if (ws_ && ws_->isConnected()) {
        sendMessage(SignalingProtocol::buildICECandidateMessage(target_peer, candidate));
    }
}

bool SignalingClient::isConnected() const {
    return ws_ && ws_->isConnected();
}

void SignalingClient::connectionDidEstablish(Connection& /*connection*/) {
    LOG_INFO("[Signaling] Connected to server");

    // Reset reconnection state on successful connect
    current_reconnect_delay_ms_ = config_.reconnect_delay_ms;
    reconnect_attempts_ = 0;

    setState(SignalingClientState::CONNECTED);

    // Automatically join room if we have a room ID
    if (!room_id_.empty()) {
        sendJoin();
    }
}

void SignalingClient::connectionDidReceive(Connection& /*connection*/, const Payload& payload) {
    if (payload.isText()) {
        handleMessage(payload.asString());
    }
}

void SignalingClient::connectionDidClose(Connection& /*connection*/) {
    LOG_INFO("[Signaling] Connection closed");

    const bool was_in_room = (state_ == SignalingClientState::IN_ROOM);
    setState(SignalingClientState::DISCONNECTED);

    // Attempt reconnection if appropriate
    if (was_in_room && should_reconnect_) {
        scheduleReconnect();
    }
}

void SignalingClient::connectionDidError(Connection& /*connection*/, const std::string& error) {
    LOG_ERROR("[Signaling] Error: %s", error.c_str());
    if (delegate_) {
        delegate_->signalingClientDidReceiveError(*this, error);
    }
}

void SignalingClient::sendMessage(const std::string& json) {
    if (ws_ && ws_->isConnected()) {
        ws_->send(Payload(json));
    }
}

void SignalingClient::handleMessage(const std::string& json) {
    const std::string type = SignalingProtocol::parseMessageType(json);

    if (type == SignalingMessageType::JOINED) {
        std::string room;
        std::vector<std::string> peers;
        if (SignalingProtocol::parseJoinedMessage(json, room, peers)) {
            setState(SignalingClientState::IN_ROOM);
            if (delegate_) {
                delegate_->signalingClientDidJoinRoom(*this, room, peers);
            }
        }
    } else if (type == SignalingMessageType::PEER_JOINED) {
        std::string peer_id;
        if (SignalingProtocol::parsePeerJoinedMessage(json, peer_id)) {
            if (delegate_) {
                delegate_->signalingClientPeerDidJoin(*this, peer_id);
            }
        }
    } else if (type == SignalingMessageType::PEER_LEFT) {
        std::string peer_id;
        if (SignalingProtocol::parsePeerLeftMessage(json, peer_id)) {
            if (delegate_) {
                delegate_->signalingClientPeerDidLeave(*this, peer_id);
            }
        }
    } else if (type == SignalingMessageType::OFFER) {
        std::string from_peer;
        SessionDescription sdp;
        if (SignalingProtocol::parseOfferMessage(json, from_peer, sdp)) {
            if (delegate_) {
                delegate_->signalingClientDidReceiveOffer(*this, from_peer, sdp);
            }
        }
    } else if (type == SignalingMessageType::ANSWER) {
        std::string from_peer;
        SessionDescription sdp;
        if (SignalingProtocol::parseAnswerMessage(json, from_peer, sdp)) {
            if (delegate_) {
                delegate_->signalingClientDidReceiveAnswer(*this, from_peer, sdp);
            }
        }
    } else if (type == SignalingMessageType::ICE_CANDIDATE) {
        std::string from_peer;
        ICECandidate candidate;
        if (SignalingProtocol::parseICECandidateMessage(json, from_peer, candidate)) {
            if (delegate_) {
                delegate_->signalingClientDidReceiveICECandidate(*this, from_peer, candidate);
            }
        }
    } else if (type == SignalingMessageType::PEER_LIST) {
        // Peer list message - we handle this via joined message with peers array
        // but log for now
    } else if (type == SignalingMessageType::ERROR) {
        std::string error;
        if (SignalingProtocol::parseErrorMessage(json, error)) {
            if (delegate_) {
                delegate_->signalingClientDidReceiveError(*this, error);
            }
        }
    }
}

void SignalingClient::sendJoin() {
    if (!room_id_.empty()) {
        sendMessage(SignalingProtocol::buildJoinMessage(room_id_, peer_id_));
    }
}

void SignalingClient::scheduleReconnect() {
    if (reconnect_attempts_ >= config_.max_reconnect_attempts) {
        if (delegate_) {
            delegate_->signalingClientMaxReconnectsReached(*this);
        }
        return;
    }

    setState(SignalingClientState::RECONNECTING);

    if (delegate_) {
        delegate_->signalingClientWillReconnect(*this, reconnect_attempts_ + 1,
                                                current_reconnect_delay_ms_);
    }

    // Note: Timer implementation is platform-specific
    // For now, we rely on the caller to call attemptReconnect() after the delay
    // In a real implementation, this would schedule a callback

    // Update delay with exponential backoff + jitter
    static std::random_device rd;
    static std::mt19937 gen(rd());
    std::uniform_real_distribution<> dis(0.85, 1.15);

    current_reconnect_delay_ms_ = static_cast<int>(std::min(
        static_cast<double>(current_reconnect_delay_ms_) * config_.reconnect_multiplier * dis(gen),
        static_cast<double>(config_.max_reconnect_delay_ms)));
}

void SignalingClient::attemptReconnect() {
    if (!should_reconnect_) {
        return;
    }

    ++reconnect_attempts_;

    // Re-create WebSocket and connect
    ws_ = WSConnection::make(config_.url);
    if (!ws_) {
        scheduleReconnect();
        return;
    }

    ws_->setDelegate(this);
    ws_->connect();
}

void SignalingClient::cancelReconnect() {
    // In a real implementation, this would cancel any scheduled timer
    should_reconnect_ = false;
}

void SignalingClient::setState(SignalingClientState new_state) {
    if (state_ == new_state) {
        return;
    }

    state_ = new_state;

    if (delegate_) {
        delegate_->signalingClientStateDidChange(*this, state_);
    }
}

}  // namespace cells::net
